packaging/maven_artifact.py

```diff
@@ -0,0 +1,376 @@
+
+    def chunk_report(self, bytes_so_far, chunk_size, total_size):
+        percent = float(bytes_so_far) / total_size
+        percent = round(percent * 100, 2)
+        sys.stdout.write("Downloaded %d of %d bytes (%0.2f%%)\r" %
+                         (bytes_so_far, total_size, percent))
+
+        if bytes_so_far >= total_size:
+            sys.stdout.write('\n')
+
+    def _write_chunks(self, response, file, chunk_size=8192, report_hook=None):
+        total_size = response.info().getheader('Content-Length').strip()
+        total_size = int(total_size)
+        bytes_so_far = 0
+
+        while 1:
+            chunk = response.read(chunk_size)
+            bytes_so_far += len(chunk)
+
+            if not chunk:
+                break
+
+            file.write(chunk)
+            if report_hook:
+                report_hook(bytes_so_far, chunk_size, total_size)
+
+        return bytes_so_far
+
+    def verify_md5(self, file, remote_md5):
+        if not os.path.exists(file):
+            return False
+        else:
+            local_md5 = self._local_md5(file)
+            remote = self._request(remote_md5, "Failed to download MD5", lambda r: r.read())
+            return local_md5 == remote
+
+    def _local_md5(self, file):
+        md5 = hashlib.md5()
+        with open(file, 'rb') as f:
+            for chunk in iter(lambda: f.read(8192), ''):
+                md5.update(chunk)
+        return md5.hexdigest()
+
+
+def main():
+    module = AnsibleModule(
+        argument_spec = dict(
+            group_id = dict(default=None),
+            artifact_id = dict(default=None),
+            version = dict(default=None),
+            classifier = dict(default=None),
+            extension = dict(default=None),
+            repository_url = dict(default=None),
+            username = dict(default=None),
+            password = dict(default=None),
+            state = dict(default="present", choices=["present","absent"]), # TODO - Implement a "latest" state 
+            dest = dict(default=None),
+        )
+    )
+
+    group_id = module.params["group_id"]
+    artifact_id = module.params["artifact_id"]
+    version = module.params["version"]
+    classifier = module.params["classifier"]
+    extension = module.params["extension"]
+    repository_url = module.params["repository_url"]
+    repository_username = module.params["username"]
+    repository_password = module.params["password"]
+    state = module.params["state"]
+    dest = module.params["dest"]
+
+    if not repository_url:
+        repository_url = "http://repo1.maven.org/maven2"
+
+    downloader = MavenDownloader(repository_url, repository_username, repository_password)
+
+    try:
+        artifact = Artifact(group_id, artifact_id, version, classifier, extension)
+    except ValueError as e:
+        module.fail_json(msg=e.args[0])
+
+    prev_state = "absent"
+    if os.path.isdir(dest):
+        dest = dest + "/" + artifact_id + "-" + version + ".jar"
```

Comment:
If a specific version is not provided, this path creates a runtime error since version = NoneType. Also, the requested extension is not taken into account.